### DISCUSSION TITLE
Can we have auto-fitting/scaling text?

### DISCUSSION THREAD

**Discussion by HubKing** [tomasklaen/uosc#542] [Q&A] [answered]
The default UI's text automatically gets scaled depending on the window size. This is good when the window is large, but bad when the window is small. It seems that in UOSC, the text size is fixed. This is good when the window is small, but bad when the window is large.

I wonder if we can get the best of both worlds: auto-scaling text, but with minimum font size limitation (that is, the text would not become illegibly small when the window is tiny).

**Comment by tomasklaen** [comment, marked as answer]
> bad when the window is large

Why? I personally don't like oversized UI elements and how default UI scales everything. uosc is designed for everything to be readable and usable on all window sizes. There are already `*_fullscreen` suffixed options for increasing sizes of elements in fullscreen or maximized mode, and I think that covers most needs.

**Reply by HubKing** [reply]
> Why? I personally don't like oversized UI elements 

Yeah, of course, you can personally dislike it, but people are different, and I like big, easy-to-click UI. There already is UI scaling option, so if the feature currently does not exist, all it needs is to applying a UI scaling value relative to the screen size when the screen size changes, unless the calculated UI scaling value is too small. 

Well, if there aren't many people who wants that, I guess I could see if I could add such feature myself for my own use.

**Reply by christoph-heinrich** [reply]
> I guess I could see if I could add such feature myself for my own use.

The simplest way would be to cap the resolution. Make sure that the aspect ratio stays correct though. A side effect of that will be that some lines won't be as crisp anymore because it's not using the native resolution for drawing things (the same thing that currently happens with not integer hidpi scaling factors).

**Reply by tomasklaen** [reply]
> I like big, easy-to-click UI

If you resize a browser window, do you also want it's UI to scale with the window, or do you want to enjoy more space for both the content and UI elements?

Currently, if control bar doesn't have enough horizontal space to accommodate all controls, it starts hiding those in the middle, and if you want to see them, you can just scale the window up a bit. If UI scaled and you lets say opened a video with portrait aspect ratio, you'd never see them, and would have to go fullscreen to use them.

To return to the original claim:

> The default UI's text automatically gets scaled depending on the window size. This is good when the window is large, but bad when the window is small

It's bad in both scenarios. Default osc is there just to provide _some_ UI for a primarily headless player, and it's mostly an after thought. As such it is implemented in a bit of a lazy way where it just scales with window size, because then you don't have to care about elements overlapping, overflowing, don't have to make things responsive, or implement any content flow. It's just good enough to be able to provide basic features if you absolutely **have** to use it.

This means it is unusable when window is narrow, and has 300px big buttons (I'm exaggerating) when window is wide. Both of these make no sense to me.

Is there any other player or an app where UI scales with window size? This has always felt like an mpv specific limitation due to not enough resources or interest in creating a UI, and I'm surprised someone would request it as a feature.

**Reply by christoph-heinrich** [reply]
With the `vidscale=no` config the original osd has a fixed size independent of window size.
It's code can be hard to deal with though. Supporting multiple UI styles certainly didn't help in that regard.

**Reply by HubKing** [reply]
> If you resize a browser window

The UI of a web browser is always displayed, taking up the screen space. However, the UI of MPV is only there when I need it, and gets hidden when i don't need it. Therefore, I don't want the UI of the web browser to become big, but I am alright with MPV's UI getting big.

> Is there any other player or an app where UI scales with window size? This has always felt like an mpv specific limitation due to not enough resources or interest in creating a UI, and I'm surprised someone would request it as a feature.

I just tested VLC and its on-screen text messages like volume change or subtitle were proportional to the window size. Before switching to Linux, I had been using [PotPlayer](https://en.wikipedia.org/wiki/PotPlayer), and if I remember it correctly, its on-screen text messages were also proportional to the window size. Its control UI was not resized, though, and I often felt it difficult to click the small control buttons. The problem of too-small on-screen text did not occur in PotPlayer because it had minimum window size limitation (the user cannot make the window smaller than a certain size)

**Reply by hooke007** [reply]
> Is there any other player or an app where UI scales with window size? This has always felt like an mpv specific limitation due to not enough resources or interest in creating a UI, and I'm surprised someone would request it as a feature.

If we are talking real GUI , there is no frontend supporting this Weird scaling. And I think would never.
If dummy ui such as OSD/osc things are counted, it's common.

>  but bad when the window is small.

It's up to usages. I often use a PIP-like mode, which means mpv's window is a tenth of the size of display (or even smaller). The vanilla osc is almost unusable because of scaling with window. The text is too small to read and the element is too little to click.
